x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java

```diff
@@ -0,0 +1,258 @@
+            t -> t.getParams() instanceof SnapshotUpgradeTaskParams
+                && ((SnapshotUpgradeTaskParams)t.getParams()).getJobId().equals(request.getJobId())).isEmpty() == false)) {
+            listener.onFailure(ExceptionsHelper.conflictStatusException(
+                "Cannot upgrade job [{}] snapshot [{}] as there is currently a snapshot for this job being upgraded",
+                request.getJobId(),
+                request.getSnapshotId()));
+            return;
+        }
+
+        final SnapshotUpgradeTaskParams params = new SnapshotUpgradeTaskParams(request.getJobId(), request.getSnapshotId());
+        // Wait for job to be started
+        ActionListener<PersistentTask<SnapshotUpgradeTaskParams>> waitForJobToStart = ActionListener.wrap(
+            persistentTask -> waitForJobStarted(persistentTask.getId(), params, request, listener),
+            e -> {
+                if (ExceptionsHelper.unwrapCause(e) instanceof ResourceAlreadyExistsException) {
+                    e = ExceptionsHelper.conflictStatusException(
+                        "Cannot upgrade job [{}] snapshot [{}] because upgrade is already in progress",
+                        e,
+                        request.getJobId(),
+                        request.getSnapshotId());
+                }
+                listener.onFailure(e);
+            });
+
+        // Start job task
+        ActionListener<Long> memoryRequirementRefreshListener = ActionListener.wrap(
+            mem -> {
+                logger.info("[{}] [{}] sending start upgrade request", params.getJobId(), params.getSnapshotId());
+                persistentTasksService.sendStartRequest(
+                    MlTasks.snapshotUpgradeTaskId(params.getJobId(), params.getSnapshotId()),
+                    MlTasks.JOB_SNAPSHOT_UPGRADE_TASK_NAME,
+                    params,
+                    waitForJobToStart);
+            },
+            listener::onFailure
+        );
+
+        // Check that model snapshot exists and should actually be upgraded
+        // Then refresh the memory
+        ActionListener<Result<ModelSnapshot>> getSnapshotHandler = ActionListener.wrap(
+            response -> {
+                if (response == null) {
+                    listener.onFailure(
+                        new ResourceNotFoundException(
+                            Messages.getMessage(Messages.REST_NO_SUCH_MODEL_SNAPSHOT, request.getSnapshotId(), request.getJobId())));
+                    return;
+                }
+                if (response.result.getMinVersion().major >= UPGRADE_FROM_MAJOR) {
+                    listener.onFailure(ExceptionsHelper.conflictStatusException(
+                        "Cannot upgrade job [{}] snapshot [{}] as it is already compatible with current major version {}",
+                        request.getJobId(),
+                        request.getSnapshotId(),
+                        UPGRADE_FROM_MAJOR));
+                    return;
+                }
```

Comment:
I opted for a failure here instead of a silent OK as it seemed nicer to say "hey, we didn't do anything" than that the update was completed.

Especially since this API should only be used when a deprecation warning indicates that a snapshot is too old to use in the next major version.

Reply:
I would do the check on the exact version rather than just the major.  Although it shouldn't be necessary, upgrading the format from e.g. 7.0 format to 7.11 format might be a useful piece of functionality to have in the future to work around some other bug.

Reply:
I can do that. It just didn't seem 100% necessary given the current state of things. Also, it doesn't seem that the `min_version` is really ever changed to the latest version. So, it very well could be that somebody calls this api numerous times in a row and since `min_version` is never changed, no error is returned. 

I suppose this is "buyer beware" and the API won't hurt anything directly.